odiglet/odiglet.go

```diff
@@ -168,35 +173,111 @@ func (o *Odiglet) Run(ctx context.Context) {
 
 func OdigletInitPhase(clientset *kubernetes.Clientset) {
 	odigletInitPhaseStart := time.Now()
+	defer func() {
+		log.Logger.V(0).Info("Odiglet init phase finished", "duration", time.Since(odigletInitPhaseStart))
+	}()
+
+	// Initialize logging
 	if err := log.Init(); err != nil {
 		panic(err)
 	}
-	err := fs.CopyAgentsDirectoryToHost()
-	if err != nil {
+
+	// Step 1: Copy instrumentation agents to host
+	if err := copyAgentsToHost(); err != nil {
 		log.Logger.Error(err, "Failed to copy agents directory to host")
 		os.Exit(-1)
 	}
 
-	nn, ok := os.LookupEnv(k8sconsts.NodeNameEnvVar)
+	// Step 2: Prepare node for Odigos installation (labels, taints)
+	if err := prepareNode(clientset); err != nil {
+		log.Logger.Error(err, "Failed to prepare node for Odigos installation")
+		os.Exit(-1)
+	}
+
+	// Step 3: Apply SELinux settings (must be last - uses chroot)
+	if err := applySecuritySettings(); err != nil {
+		log.Logger.Error(err, "Failed to apply SELinux settings on RHEL host")
+		os.Exit(-1)
+	}
+
+	os.Exit(0)
+}
+
+// copyAgentsToHost copies instrumentation agents from the odiglet image to the host filesystem.
+func copyAgentsToHost() error {
+	return fs.CopyAgentsDirectoryToHost()
+}
+
+// prepareNode prepares the Kubernetes node for Odigos installation by updating labels and taints.
+func prepareNode(clientset *kubernetes.Clientset) error {
+	nodeName, ok := os.LookupEnv(k8sconsts.NodeNameEnvVar)
+	if !ok {
+		return fmt.Errorf("env var %s is not set", k8sconsts.NodeNameEnvVar)
+	}
+
+	_, err := k8snode.PrepareNodeForOdigosInstallation(clientset, nodeName)
+	if err != nil {
+		return err
+	}
+
+	log.Logger.Info("Successfully prepared node for Odigos installation")
+	return nil
+}
+
+// applySecuritySettings applies security settings like SELinux contexts.
+// This must be called last as it uses chroot to access host commands.
+func applySecuritySettings() error {
+	return fs.ApplyOpenShiftSELinuxSettings()
+}
+
+// OdigletDiscoveryPhase runs the node discovery phase to collect and persist node details.
+// This is run as a separate container/mode to collect node characteristics and capabilities.
+// It checks all registered features (OSS + enterprise extensions) and creates a NodeDetails CRD.
+func OdigletDiscoveryPhase(config *rest.Config, clientset *kubernetes.Clientset) {
+	discoveryPhaseStart := time.Now()
+	defer func() {
+		log.Logger.V(0).Info("Odiglet discovery phase finished", "duration", time.Since(discoveryPhaseStart))
+	}()
+
+	// Initialize logging
+	if err := log.Init(); err != nil {
+		panic(err)
+	}
```

Comment:
should this be in the `OdigletDiscoveryPhase` function? isn't it something we want to do in `main.go`? or somewhere else but not in the discovery phase?